### PR TITLE
added equality comparison for Position objects

### PR DESCRIPTION
**Pull request by peteraldous**
`Parsers.Parser.|||` was not working as expected. I looked into it and discovered that `Position` objects were using pointer equality, not structural equality. I added this equality function and it fixed the issue.

I won't be offended if there are changes that need to be made to make this more idiomatic or whatever.

### CODE REVIEW

**Comment by SethTisue** [comment]
> Parsers.Parser.||| was not working as expected

perhaps you could add a test case to demonstrate the issue that is now fixed...?

File: shared/src/main/scala/scala/util/parsing/input/Position.scala [hunk 1]
```diff
@@ -60,4 +60,11 @@ trait Position {
     this.line < that.line ||
     this.line == that.line && this.column < that.column
   }
+
+  /** Compare this position to another, checking for equality.
+   *
+   * @param `that` a `Position` to compare to this `Position`
+   * @return true if the line numbers and column numbers are equal.
+   */
+  def ==(that: Position) = this.line == that.line && this.column == that.column
```
Comment (SethTisue) [inline shared/src/main/scala/scala/util/parsing/input/Position.scala:69]: it's a peculiarity of Scala that at call sites one is supposed to use `==`, but at the definition site one is supposed to `override def equals`, not `def ==`

**Comment by SethTisue** [comment]
It's a bit puzzling to me that this would have fixed anything, given that the only two subtypes of `Position` that I see are `NoPosition` and `OffsetPosition`, and the former is a singleton for which reference equality should be fine, and the latter is a `case class` which should have an automatically generated `equals`. Though in the latter it's comparing `offset` instead of `line` and `column`, as in your change — is the latter somehow more correct?

**Comment by peteraldous** [comment]
I'm not seeing the same behavior when I run `sbt test` on my machine. Is there a way I can configure my machine so it produces the same error as the workflow above?

**Comment by Philippus** [comment]
> I'm not seeing the same behavior when I run `sbt test` on my machine. Is there a way I can configure my machine so it produces the same error as the workflow above?

I'm also having trouble with triggering the same error locally.

**Comment by peteraldous** [comment]
The puzzle might be explained by the fact that https://github.com/scala/scala-parser-combinators/blob/main/shared/src/main/scala/scala/util/parsing/input/StreamReader.scala#L69 creates an _anonymous_ subclass of `Position`.

**Comment by peteraldous** [comment]
Thanks for the response. I have plenty on my plate, as well, so I understand.

I am happy to further defend the change but thought I'd answered all of your questions: the tests I added demonstrate correct behavior and if you comment out the `equals` method I added, the last of the three breaks. I also added a comment with a pointer to code that makes an anonymous subclass of `Position` that doesn't have an `equals` method.

I would much rather that you wait until you're comfortable merging. Is there anything else I can answer or tweak? (Happy to have comments from others on the repo, too.)

**Comment by SethTisue** [comment]
Thank you!

**Pull request merged**